src/nlv_object.h

```diff
@@ -5,182 +5,184 @@
-    ref_ = NULL;
-    valid_ = false;
   }
 
-protected:
-  NLVObjectBase *ref_;
-  bool valid_;
-};
-
-#define NLV_STRINGIFY0(v) #v
-#define NLV_STRINGIFY(v) NLV_STRINGIFY0(v)
-
-// macro could be removed but alternate solution seems overly complex
-// https://blog.molecular-matters.com/2015/12/11/getting-the-type-of-a-template-argument-as-string-without-rtti/
-#define NLV_OBJECT_STATIC_HELPERS(class_name) \
-  static const char* ClassName() { \
-    return NLV_STRINGIFY(class_name); \
-  } \
-  friend class NLVObject;
+  template <typename ParentClass, typename HandleType, typename CleanupHandler>
+  class NLVObject : public NLVObjectBase
+  {
+  public:
+    typedef HandleType handle_type;
 
+    typedef typename std::remove_pointer<HandleType>::type HandleValue;
 
-class NLVObjectBase : public Nan::ObjectWrap
-{
-public:
-  virtual void ClearHandle() = 0;
-  virtual void ClearChildren() = 0;
-  virtual void SetParentReference(NLVObjectBasePtr *parentReference) = 0;
+    NLVObject(HandleType handle) : handle_(handle, CleanupHandler::cleanup), parentReference_(NULL) {}
+    ~NLVObject() {
+      // calling virtual ClearHandle() will break if overridden by subclasses
+      ClearHandle();
+      if (parentReference_ != NULL) {
+        parentReference_->SetInvalid();
+      }
+    }
 
-  std::vector<NLVObjectBasePtr*> children_;
-};
+    struct MethodNoReturnHandler {
+      template<typename T>
+      Worker::OnFinishedHandler operator()(T val) {
+        return NLV::PrimitiveReturnHandler(true);
+      }
+    };
 
-template <typename ParentClass, typename HandleType, typename CleanupHandler>
-class NLVObject : public NLVObjectBase
-{
-public:
-  typedef HandleType handle_type;
+    template<typename T>
+    struct MethodReturn {
+      Worker::OnFinishedHandler operator()(T val) {
+        return NLV::PrimitiveReturnHandler(static_cast<T>(val));
+      }
+    };
+    using MethodReturnBool = MethodReturn<bool>;
+    using MethodReturnInt = MethodReturn<int>;
+
+    struct MethodReturnString {
+      Worker::OnFinishedHandler operator()(const char* val) {
+        std::string str = val;
+        return [=](Worker* worker) {
+          Nan::HandleScope scope;
+          v8::Local<v8::Value> argv[] = { Nan::Null(), Nan::New(str).ToLocalChecked() };
+          worker->Call(2, argv);
+        };
+      }
+    };
+    
+    static inline bool virHasFailed(int val) {
+      return val < 0;
+    }
+    
+    static inline bool virHasFailed(const void* ptr) {
+      return ptr == nullptr;
+    }
 
-  typedef typename std::remove_pointer<HandleType>::type HandleValue;
+    template<typename ReturnHandler = MethodNoReturnHandler, typename Z, typename... Args>
+    static void RunMethod(const Nan::FunctionCallbackInfo<v8::Value>& info, Z virFunction, Args... args) {
+      Nan::HandleScope scope;
+      virDomainPtr handle = ParentClass::Unwrap(info.This())->virHandle();
+      auto virCaller = std::bind(virFunction, handle, args...);
```

Comment:
It's due to GCC 4.8 bug https://gcc.gnu.org/bugzilla/show_bug.cgi?id=55914#c3

I need to add a comment for that.